src/perception/cost_map/cost_map.processing.cpp

```diff
@@ -0,0 +1,54 @@
+#include "cost_map.hpp"
+
+namespace mrover {
+
+    auto CostMapNodelet::pointCloudCallback(sensor_msgs::PointCloud2ConstPtr const& msg) -> void {
+        assert(msg);
+        assert(msg->height > 0);
+        assert(msg->width > 0);
+
+        if (!mPublishCostMap) return;
+
+        try {
+            SE3d zed_to_map = SE3Conversions::fromTfTree(mTfBuffer, "map", "zed_left_camera_frame");
+            auto* points = reinterpret_cast<Point const*>(msg->data.data());
+            // std::for_each(points, points + msg->width * msg->height, [&](auto* point) {
+            for (auto point = points; point - points < msg->width * msg->height; point += mDownSamplingFactor) {
+                Eigen::Vector4d p{point->x, point->y, point->z, 1};
+                point_matrix.col((point - points) / mDownSamplingFactor) = p;
+                Eigen::Vector4d normal{point->normal_x, point->normal_y, point->normal_z, 0};
+                normal_matrix.col((point - points) / mDownSamplingFactor) = normal;
+            }
+
+            // TODO(neven): Make sure this still works with manif
+            point_matrix = zed_to_map.transform() * point_matrix;
+            normal_matrix = zed_to_map.transform() * normal_matrix;
+
+            for (uint32_t i = 0; i < mNumPoints; i++) {
+                double x = point_matrix(0, i);
+                double y = point_matrix(1, i);
+                Eigen::Vector4d n = normal_matrix.col(i);
```

Comment:
`Eigen::Vector3d normalInMap = normal_matrix.col(i).head<3>()`

`head` will grab the first three. No need to do `Eigen::Vector3d normal{n.x(), n.y(), n.z()};` later.

This automatically de-homogenizes the vector. Note that we are not dividing by w which is usually done. But we know w=1 because we are multipliying by a SO3 matrix which is orthnormal and has det=1 so w won't change. Distance between points in your rigidbody is preservedi n other words